### PR TITLE
Add a feature that offers to add missing accessibility modifiers.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @Pilchie

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 1]
```diff
@@ -1271,12 +1271,16 @@ This version used in: {2}</value>
   <data name="in_0_project_1" xml:space="preserve">
     <value>in {0} (project {1})</value>
   </data>
+<<<<<<< HEAD
```
Comment (Therzok) [inline src/Features/Core/Portable/FeaturesResources.resx]: Merge conflict marker

File: src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+            // Have to have a name to report the issue on.
+            var name = member.GetNameToken();
+            if (name.Kind() == SyntaxKind.None)
+            {
+                return;
+            }
+
+            // If they already have accessibility, no need to report anything.
+            var accessibility = generator.GetAccessibility(member);
```
Comment (Therzok) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: Is there a editorconfig option that needs to be supported here? i.e. `avoid usage of implicit modifiers` (private for members/nested classes, internal for top-level classes)

Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: Yes, an editor config option was added for this "..._require_accessibility_modifiers"

File: src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            if (accessibility != Accessibility.NotApplicable)
+            {
+                return;
+            }
+
+            // Certain members never have accessibility. Don't bother reporting on them.
+
+            if (member.IsKind(SyntaxKind.DestructorDeclaration))
```
Comment (Therzok) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: `generator.GetAccessibility` will already report NotApplicable for destructor declarations.

File: src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            {
+                return;
+            }
+
+            if (member.IsKind(SyntaxKind.ConstructorDeclaration, out ConstructorDeclarationSyntax constructor) &&
+                constructor.Modifiers.Any(SyntaxKind.StaticKeyword))
```
Comment (Therzok) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: Move this static modifier check into [CanHaveAccessibility](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/CodeGeneration/CSharpSyntaxGenerator.cs,1411) ?

File: src/Features/Core/Portable/AddAccessibilityModifiers/AbstractAddAccessibilityModifiersCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                editor.ReplaceNode(
+                    declaration,
+                    (currentDeclaration, generator) =>
+                    {
+                        return generator.WithAccessibility(currentDeclaration, symbol.DeclaredAccessibility);
+                    });
+            }
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
```
Comment (Therzok) [inline src/Features/Core/Portable/AddAccessibilityModifiers/AbstractAddAccessibilityModifiersCodeFixProvider.cs]: `AddAccessibilityCodeAction`?

File: src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs (hunk 2 above)
Comment (Therzok) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: Oh, it's done in the code fix provider. My bad.

File: src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs (hunk 3 above)
Comment (Therzok) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: Okay, it seems like `NotApplicable` means that no accessibility is specified, not that the node cannot take accessibility modifiers.

Comment (Therzok) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: Having NotApplicable mean both `node cannot take modifiers` from CanHaveAccessibility evaluating to false and `no modifiers on node` from `GetAccessibilityAndModifiers` is confusing. Maybe that should be fixed too?

File: src/EditorFeatures/VisualBasicTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.vb [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        End Function
+
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddAccessibilityModifiers)>
+        Public Async Function TestAllConstructs() As Task
+            Await TestInRegularAndScriptAsync(
+"
+namespace N
+    namespace Outer.Inner
+        class {|FixAllInDocument:C|}
+            class NestedClass
```
Comment (jasonmalinowski) [inline src/EditorFeatures/VisualBasicTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.vb]: Is this all lowercase intentionally?

File: src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            }
+
+            // If we have a class or struct, recurse inwards.
+            if (member.IsKind(SyntaxKind.ClassDeclaration, out TypeDeclarationSyntax typeDeclaration) ||
+                member.IsKind(SyntaxKind.StructDeclaration, out typeDeclaration))
+            {
+                ProcessMembers(context, generator, option, typeDeclaration.Members);
+            }
+
+#if false
```
Comment (jasonmalinowski) [inline src/Features/CSharp/Portable/AddAccessibilityModifiers/CSharpAddAccessibilityModifiersDiagnosticAnalyzer.cs]: Are we tracking this via a bug somewhere in a GitHub project or the like?

File: src/EditorFeatures/CSharpTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            event Action I.e6 { add { } remote { } }
+
+            static C() { }
+            C() { }
+            public C(int i) { }
+
+            ~C() { }
+
+            void M1() { }
+            public void M2() { }
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.cs]: Consider: partial methods.

File: src/Features/VisualBasic/Portable/AddAccessibilityModifiers/VisualBasicAddAccessibilityModifiersDiagnosticAnalyzer.vb [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+
+            If member.Kind() = SyntaxKind.NamespaceBlock Then
+                Dim namespaceBlock = DirectCast(member, NamespaceBlockSyntax)
+                ProcessMembers(context, generator, [option], namespaceBlock.Members)
+            End If
+
+            ' If we have a class or struct or module, recurse inwards.
+            If member.IsKind(SyntaxKind.ClassBlock) OrElse
+           member.IsKind(SyntaxKind.StructureBlock) OrElse
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/AddAccessibilityModifiers/VisualBasicAddAccessibilityModifiersDiagnosticAnalyzer.vb]: Strange indenting is strange.

File: src/Features/VisualBasic/Portable/AddAccessibilityModifiers/VisualBasicAddAccessibilityModifiersDiagnosticAnalyzer.vb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+            ' If they already have accessibility, no need to report anything.
+            Dim Accessibility = generator.GetAccessibility(member)
+            If Accessibility <> Accessibility.NotApplicable Then
+                Return
+            End If
+
+            ' Missing accessibility.  Report issue to user.
+            Dim additionalLocations = ImmutableArray.Create(member.GetLocation())
+            context.ReportDiagnostic(Diagnostic.Create(
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/AddAccessibilityModifiers/VisualBasicAddAccessibilityModifiersDiagnosticAnalyzer.vb]: Indenting

File: src/Workspaces/Core/Portable/CodeStyle/AccessibilityModifiersRequired.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    internal enum AccessibilityModifiersRequired
+    {
+        Never = 0,
+        Always = 1,
+
+        // Future proofing for when C# adds default interface methods.  At that point
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/CodeStyle/AccessibilityModifiersRequired.cs]: Tag a bug in the GitHub project so we can find this later?

File: src/Workspaces/Core/Portable/CodeStyle/CodeStyleOptions.cs [hunk 12]
```diff
@@ -132,5 +133,43 @@ public class CodeStyleOptions
...
+            new PerLanguageOption<CodeStyleOption<AccessibilityModifiersRequired>>(
+                nameof(CodeStyleOptions), nameof(RequireAccessibilityModifiers), defaultValue: s_requireAccessibilityModifiersDefault,
+                storageLocations: new OptionStorageLocation[]{
+                    new EditorConfigStorageLocation<CodeStyleOption<AccessibilityModifiersRequired>>("dotnet_style_require_accessibility_modifiers", s => ParseAccessibilityModifiersRequired(s)),
+                    new RoamingProfileStorageLocation("TextEditor.%LANGUAGE%.Specific.RequireAccessibilityModifiers")});
+
+        private static CodeStyleOption<AccessibilityModifiersRequired> ParseAccessibilityModifiersRequired(string optionString)
+        {
+            if (TryGetCodeStyleValueAndOptionalNotification(optionString,
+                    out var value, out var notificationOpt))
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/CodeStyle/CodeStyleOptions.cs]: Indenting.

File: src/EditorFeatures/VisualBasicTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.vb (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.vb]: just how i code VB :)

File: src/EditorFeatures/CSharpTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.cs (hunk 8 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/AddAccessibilityModifiers/AddAccessibilityModifiersTests.cs]: Added support for this.

File: src/Features/VisualBasic/Portable/AddAccessibilityModifiers/VisualBasicAddAccessibilityModifiersDiagnosticAnalyzer.vb (hunk 9 above)
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/AddAccessibilityModifiers/VisualBasicAddAccessibilityModifiersDiagnosticAnalyzer.vb]: fixed.

File: src/Workspaces/Core/Portable/CodeStyle/CodeStyleOptions.cs (hunk 12 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeStyle/CodeStyleOptions.cs]: intentiona.

**Comment by CyrusNajmabadi** [comment]
@kuhlenh New feature.

**Pull request merged**